dask/utils.py

```diff
@@ -654,3 +660,61 @@ def name_function(i):
         return str(i).zfill(pad_length)
 
     return name_function
+
+
+def infer_storage_options(urlpath, inherit_storage_options=None):
+    """ Infer storage options from URL path and merge it with existing storage
+    options.
+
+    Parameters
+    ----------
+    urlpath: str or unicode
+        Either local absolute file path or URL (hdfs://namenode:8020/file.csv)
+    storage_options: dict (optional)
+        Its contents will get merged with the infered information from the
+        given path
+
+    Returns
+    -------
+    Storage options dict.
+
+    Examples
+    --------
+    >>> infer_storage_options('/mnt/datasets/test.csv')  # doctest: +SKIP
+    {"_protocol": "file", "_path", "/mnt/datasets/test.csv"}
+    >>> infer_storage_options(
+    ...          'hdfs://username:pwd@node:123/mnt/datasets/test.csv?q=1',
+    ...          inherit_storage_options={'extra': 'value'})  # doctest: +SKIP
+    {"_protocol": "hdfs", "_username": "username", "_password": "pwd",
+    "_host": "node", "_port": 123, "_path": "/mnt/datasets/test.csv",
+    "_url_query": "q=1", "extra": "value"}
+    """
+    parsed_path = urlsplit(urlpath)
+
+    inferred_storage_options = {
+        '_protocol': parsed_path.scheme or 'file',
+        '_path': parsed_path.path,
+    }
+
+    if parsed_path.netloc:
+        inferred_storage_options['_host'] = parsed_path.hostname
+        if parsed_path.port:
+            inferred_storage_options['_port'] = parsed_path.port
+        if parsed_path.username:
+            inferred_storage_options['_username'] = parsed_path.username
+        if parsed_path.password:
+            inferred_storage_options['_password'] = parsed_path.password
+
+    if parsed_path.query:
+        inferred_storage_options['_url_query'] = parsed_path.query
+    if parsed_path.fragment:
+        inferred_storage_options['_url_fragment'] = parsed_path.fragment
```

Comment:
We probably shouldn't deviate from the standard names used in `urlsplit` if we can avoid it.  Perhaps we should use the names explicitly to avoid having to filter them out?